server/src/main/java/org/apache/druid/server/coordinator/duty/CompactSegments.java

```diff
@@ -464,6 +493,27 @@ private int submitCompactionTasks(
         }
       }
 
+      final CompactionEngine compactionEngine = config.getEngine() == null ? defaultEngine : config.getEngine();
+      final Map<String, Object> autoCompactionContext = newAutoCompactionContext(config.getTaskContext());
+      int slotsRequiredForCurrentTask;
+
+      if (compactionEngine == CompactionEngine.MSQ) {
+        if (autoCompactionContext.containsKey(ClientMsqContext.CTX_MAX_NUM_TASKS)) {
+          slotsRequiredForCurrentTask = (int) autoCompactionContext.get(ClientMsqContext.CTX_MAX_NUM_TASKS);
+        } else {
+          // Since MSQ needs all task slots for the calculated #tasks to be available upfront, allot all available
+          // compaction slots (upto a max of MAX_TASK_SLOTS_FOR_MSQ_COMPACTION) to current compaction task to avoid
+          // stalling. Setting "taskAssignment" to "auto" has the problem of not being able to determine the actual
+          // count, which is required for subsequent tasks.
+          slotsRequiredForCurrentTask = Math.min(numAvailableCompactionTaskSlots,
+                                                 MAX_TASK_SLOTS_FOR_MSQ_COMPACTION_TASK
+          );
```

Comment:
I think this can create trouble. If `numAvailableCompactionTaskSlots` is 1, then we put `maxNumTasks` as 1 in the context which causes the `CompactionTask` to fail with this exception:

```
org.apache.druid.error.DruidException: Compaction spec not supported. Reason[MSQ context maxNumTasks [1] cannot be less than 2, since at least 1 controller and 1 worker is necessary.].
```

Please address this.

Reply:
Yeah I was aware about this and was hoping that the user would override the context in cases where effective slots are insufficient. But since a local small cluster is a common case, I'm now considering overriding `slotsRequiredForCurrentTask` slots to 2 if `numAvailableCompactionTaskSlots` is 1.

Reply:
Yeah, let's do that for now.